From 4e4ad7c1cac6955ffb1d377943429c3560058109 Mon Sep 17 00:00:00 2001
From: Zoltan HERPAI <wigyori@uid0.hu>
Date: Thu, 10 Aug 2023 23:42:13 +0200
Subject: drivers: include cpufreq/Kconfig for riscv

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>
---
 drivers/Kconfig | 2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -239,6 +239,8 @@ source "drivers/most/Kconfig"
 
 source "drivers/peci/Kconfig"
 
+source "drivers/cpufreq/Kconfig"
+
 source "drivers/hte/Kconfig"
 
 source "drivers/cdx/Kconfig"
