--- a/net.c
+++ b/net.c
@@ -46,11 +46,7 @@
 #include <arpa/inet.h>
 #include <net/if.h>
 #include <asm/types.h>
-#if defined(__GLIBC__)
-# include <netipx/ipx.h>
-#else
-# include <linux/ipx.h>
-#endif
+#include <linux/ipx.h>
 
 #if defined(HAVE_LINUX_NETLINK_H)
 # include <linux/netlink.h>
--- a/time.c
+++ b/time.c
@@ -656,7 +656,7 @@ printsigevent(struct tcb *tcp, long arg)
 			/* _pad[0] is the _tid field which might not be
 			   present in the userlevel definition of the
 			   struct.  */
-			tprintf("{%d}", sev._sigev_un._pad[0]);
+			tprintf("{%d}", *(int *) &sev.sigev_notify_function);
 #elif defined(HAVE_STRUCT_SIGEVENT___PAD)
 			tprintf("{%d}", sev.__pad[0]);
 #else
--- a/evdev.c
+++ b/evdev.c
@@ -28,6 +28,8 @@
 
 #include "defs.h"
 
+#include <linux/ioctl.h>
+
 #ifdef HAVE_LINUX_INPUT_H
 #include <linux/input.h>
 #include "xlat/evdev_abs.h"
--- a/mknod.c
+++ b/mknod.c
@@ -1,6 +1,7 @@
 #include "defs.h"
 
 #include <fcntl.h>
+#include <sys/stat.h>
 
 #ifdef MAJOR_IN_SYSMACROS
 # include <sys/sysmacros.h>
--- a/printmode.c
+++ b/printmode.c
@@ -1,6 +1,7 @@
 #include "defs.h"
 
 #include <fcntl.h>
+#include <sys/stat.h>
 
 #include "xlat/modetypes.h"
 
